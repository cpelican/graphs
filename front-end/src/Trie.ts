type MapLetterNodeValue = {children: MapLetterNode, word?: string}
type MapLetterNode = Map<string, MapLetterNodeValue>

export class Trie {
    private nodes: MapLetterNode = new Map();

    public getNodes(): MapLetterNode {
        return this.nodes;
    }

    public insert(
        remainingLetters: string,
        previousWord: string = '',
        childrenNodes: MapLetterNode = this.nodes,
    ): MapLetterNode {
        if (remainingLetters.length === 0) {
            return this.nodes;
        }
        let nextRemainingLetters = remainingLetters.slice(1, remainingLetters.length),
            firstLetter = remainingLetters[0],
            currentWord: string = previousWord + firstLetter,
            childNode: MapLetterNodeValue = childrenNodes.get(firstLetter) ?? {children: new Map()};

        if (nextRemainingLetters.length == 0) {
            childNode.word = currentWord;
        }
        childrenNodes.set(firstLetter, childNode);

        return this.insert(nextRemainingLetters, currentWord, childrenNodes?.get(firstLetter)?.children as MapLetterNode);
    }
    public getSuggestions(suggestion: string): Set<string> {
        let currentLetterIndex = 0,
            currentNode: MapLetterNode = this.nodes;
        // get the suggestions tree
        while(currentLetterIndex !== suggestion.length) {
            let currentLetter = suggestion[currentLetterIndex];
            currentNode = currentNode?.get(currentLetter)?.children as MapLetterNode;
            currentLetterIndex += 1;

            if (currentNode == null) {
                break;
            }
        }

        // build the suggestions
        return this.searchSuggestions(currentNode);
    }

    public searchSuggestions(nodes: MapLetterNode, suggestions: Set<string> = new Set()): Set<string> {
        if (nodes == null) {
            return suggestions;
        }
        for (let currentNode of nodes.values()) {
            if (currentNode.word != null) {
                suggestions.add(currentNode.word as string);
            }
            this.searchSuggestions(currentNode.children as MapLetterNode, suggestions);
        }

        return suggestions;
    }
}
