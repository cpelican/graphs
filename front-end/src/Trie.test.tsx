import React from 'react';
import {Trie} from './Trie';

let trie = new Trie();
test('creates correctly the trie', () => {
  trie.insert('bimbambum');

  expect(trie.getNodes()).toEqual({
      "b": {
          "i": {
              "m": {
                  "b": {
                      "a": {
                          "m": {
                              "b": {
                                  "u": {
                                      "m": {}
                                  }
                              }
                          }
                      }
                  }
              }
          }
      }
  });
});

test('adding an other word does not deletes the first', () => {
  trie.insert('bimbeta');

  expect(trie.getNodes()).toEqual({
      "b": {
          "i": {
              "m": {
                  "b": {
                      "a": {
                          "m": {
                              "b": {
                                  "u": {
                                      "m": {}
                                  }
                              }
                          }
                      },
                      "e": {
                          "t": {
                              "a": {}
                          }
                      },
                  }
              }
          }
      }
  });
});
