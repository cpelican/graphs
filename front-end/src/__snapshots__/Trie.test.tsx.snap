// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates correctly the  trie 1`] = `
Map {
  "b" => Object {
    "children": Map {
      "i" => Object {
        "children": Map {
          "m" => Object {
            "children": Map {
              "b" => Object {
                "children": Map {
                  "a" => Object {
                    "children": Map {
                      "m" => Object {
                        "children": Map {
                          "b" => Object {
                            "children": Map {
                              "u" => Object {
                                "children": Map {
                                  "m" => Object {
                                    "children": Map {},
                                    "word": "bimbambum",
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`creates correctly the  trie 2`] = `
Map {
  "b" => Object {
    "children": Map {
      "i" => Object {
        "children": Map {
          "m" => Object {
            "children": Map {
              "b" => Object {
                "children": Map {
                  "a" => Object {
                    "children": Map {
                      "m" => Object {
                        "children": Map {
                          "b" => Object {
                            "children": Map {
                              "u" => Object {
                                "children": Map {
                                  "m" => Object {
                                    "children": Map {},
                                    "word": "bimbambum",
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "e" => Object {
                    "children": Map {
                      "t" => Object {
                        "children": Map {
                          "a" => Object {
                            "children": Map {},
                            "word": "bimbeta",
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`gets the suggestions 1`] = `
Set {
  "bimbambum",
  "bimbeta",
}
`;
